### PR TITLE
Write Implementation Tests for GTK+ Tree/Table

### PR DESCRIPTION
**Pull request by DariusMontez**
Until this point, none of the implementation-layer widgets have had their own unit tests, different from the automatically generated ones. However, due to the complexity of these to widgets, GTK+ Table and Tree, and to some bugs that were accidentally introduced, they deserve their own tests.

The tests I have written may not be complete, so for those reviewing this pull request, I'd be grateful for feedback and suggestions for additional tests. Here's a summary of what I'm testing for (both widgets):

* `insert`, `change`, and `remove` methods behave correctly
* persistent `TreeIter`s (keeping track of changing rows or nodes)
* behavior of `on_select` and any data passed with it (row or node)
* tests for various internal methods.

I still need to write tests for:

* ~~`change_source`~~ (done)

The implementation classes have, of course, been changed to comply with the tests.

### CODE REVIEW

File: src/gtk/tests/widgets/test_table.py [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+        # Put data in the Table
+        self.table.table = []
+        A = self.table.data.append(None, one="A1", two="A2")
+        tree_iter = self.gtk_table.rows[A]
+        self.assertEqual(A, self.gtk_table.get_row(tree_iter))
+
+        with self.assertRaises(TypeError):
+            self.gtk_table.get_row(None)
+
+    def test_row_persisence(self):
```
Comment (phildini) [inline src/gtk/tests/widgets/test_table.py]: nit: persistence

File: src/gtk/tests/widgets/test_table.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+        B = self.table.data.append(None, one="B1", two="B2")
+
+        # Create a flag
+        succeed = False
+
+        def on_select(table, row, *kw):
+            # Make sure the right row was selected
+            self.assertEqual(row, B)
+
+            nonlocal succeed
```
Comment (phildini) [inline src/gtk/tests/widgets/test_table.py]: TIL about nonlocal. It's a bit hard to follow exactly _which_ `succeed` is being used in the assert.

File: src/gtk/toga_gtk/widgets/table.py [hunk 3]
```diff
@@ -27,27 +27,36 @@ def create(self):
...
+                return row
+
+    def set_tree_iter(self, item, tree_iter):
+        item._tree_iters = getattr(item, "_tree_iters", {})
+        item._tree_iters[self] = tree_iter
+        self.rows[item] = tree_iter
+
     def _on_select(self, selection):
         if hasattr(self.interface, "_on_select") and self.interface.on_select:
-            tree_model, impl = selection.get_selected()
```
Comment (phildini) [inline src/gtk/toga_gtk/widgets/table.py]: `impl` is a pretty common convention in toga... I'm on the fence about changing it. 

On the one hand, this is closer to what gtk is doing. On the other hand, `impl` is how toga "thinks" about it.

**Comment by DariusMontez** [comment]
I've added tests for `change_source`. Unless someone can think of another test that should be run for GTK Table/Tree, I'm ready for review!

File: src/gtk/tests/widgets/test_table.py (hunk 1 above)
Comment (DariusMontez) [reply src/gtk/tests/widgets/test_table.py]: Good catch :-)

File: src/gtk/tests/widgets/test_table.py (hunk 2 above)
Comment (DariusMontez) [reply src/gtk/tests/widgets/test_table.py]: Agreed; under normal circumstances (i.e. not using GTK event loop), I would just have the handler raise an `Succeed` Exception and then `self.assertRaises(Succeed)`, but exceptions get raised on the event loop and seem to pass unnoticed by the test runner. I am certainly open to suggestions for a more conventional or readable method.

File: src/gtk/toga_gtk/widgets/table.py (hunk 3 above)
Comment (DariusMontez) [reply src/gtk/toga_gtk/widgets/table.py]: Good point. I'm pretty sure the main reason I changed it was because I was getting confused between all the different `impl`s going on in the same file, so I renamed it to what it actually represented to make the code easier to reason about. I'll mess with the names some more and perhaps come up with a reasonable compromise.

**Comment by DariusMontez** [comment]
@phildini I thought about your comment on `impl`. Looking over my code caused my to realize that I no longer needed to store `Gtk.TreeIter`s on `Node`s or `Row`s, since I'm keeping track of them thru `self.nodes` and `self.rows`. The files actually got smaller and simpler!

Having platform-specific data attached to a platform-independent object like `Node`/`Row` gave some good reason for `impl`, but now that the implementation keeps track of its own TreeIters, `impl` is, in my mind, no longer a correct description, since `Gtk.TreeIter`s become and internal affair for the GTK implementation. Some platforms have no equivalent to the concept of TreeIter (I think Cocoa is one?); therefore, TreeIters wouldn't be an impl ("implementation") of anything, really.

Does this reasoning make sense?

**Comment by DariusMontez** [comment]
@freakboy3742 Thank you for the review! 

The reason I went with the global `dict` is that the method of attaching TreeIters to  Node `_impl`s did not provide a clear way to map from a TreeIter to its corresponding Node. This process is required in `on_select()`. 
For Table, I can just iterate through `self.interface.data` and find the Node whose `_impl` is the selected TreeIter. 
For Tree, though, I'd have to use a recursive function, which for whatever reason feels unattractive to me (perhaps because `TreeSource` lacks a `has_children()` like Node's). A later PR for TreeSource might be worthwhile if it would clean things up. One thing at time... :-)

Anyway, this recursive function would be my `get_node()`. It would, of course, work very differently than my "global dict" `get_node()` method, but ultimately the implementation needs some way to resolve a TreeIter to a Node.

I'm working on switching back to the `_impl` method, and writing the recursive function, and I'll let you know when the change is complete.

**Comment by freakboy3742** [comment]
@DariusMontez Would it be possible to get you to open upstream commits for this PR? In the process of reviewing and testing, I've got a couple of modifications, and it would be easier to just commit what I've done, rather than explain and get you to reproduce the same code :-)

File: src/gtk/toga_gtk/widgets/tree.py [hunk 4]
```diff
@@ -80,30 +52,32 @@ def change_source(self, source):
...
 
-        def append_node(parent, root=False):
-            if self.node_has_children(parent):
-                for i, child_node in enumerate(parent):
-                    self.insert(parent, i, child_node)
-                    append_node(child_node)
+        def append_node(parent):
+            for i, child_node in enumerate(parent):
+                self.insert(parent, i, child_node)
+                append_node(child_node)
```
Comment (DariusMontez) [inline src/gtk/toga_gtk/widgets/tree.py]: Perhaps this is more an issue with the data sources / Node / Row, but as it is, iteration fails when their are no children since `_roots` or `_children` initialize to None. (`iter(None) => TypeError`)

File: src/gtk/toga_gtk/widgets/tree.py [hunk 5]
```diff
@@ -27,50 +27,22 @@ def create(self):
...
-        node._impl[self] = impl
-
-    def del_impl(self, node):
-        if hasattr(node, '_impl'):
-            del node._impl[self]
-
-    def get_impl(self, node):
-        if hasattr(node, '_impl'):
-            return node._impl.get(self, None)
-
```
Comment (DariusMontez) [inline src/gtk/toga_gtk/widgets/tree.py]: I like that you dissolved these methods -- I had forgotten that these checks are only needed once when the row is first inserted.

File: src/core/toga/widgets/table.py [hunk 6]
```diff
@@ -44,6 +44,7 @@ def __init__(self, headings, id=None, style=None, data=None, accessors=None, on_
         self.headings = headings
         self._accessors = build_accessors(headings, accessors)
 
+        self._on_select = None
```
[the remark this answers is not in the document]
Comment (freakboy3742) [reply src/core/toga/widgets/table.py]: Yes. This primes the attribute value so that `on_select` property requests *always* work. This avoids the need to do `hasattr(interface, 'on_select')` later on.

File: src/gtk/toga_gtk/widgets/tree.py (hunk 5 above)
Comment (freakboy3742) [reply src/gtk/toga_gtk/widgets/tree.py]: Yeah. Same goes for all the get/set/delete methods. `_impl` *must* exist after it's been added to a tree - otherwise there's a big problem. So you don't need to protection against non-existence of the `_impl` attribute. And there's only one way to add to a tree.
